src/native/mqtt_connection.c

```diff
@@ -101,30 +101,33 @@ static struct mqtt_jni_async_callback *mqtt_jni_async_callback_new(
     struct mqtt_jni_connection *connection,
     jobject async_callback) {
 
-    JNIEnv *env = aws_jni_get_thread_env(connection->jvm);
+    /********** JNI ENV ACQUIRE **********/
+    JNIEnv *env = aws_jni_acquire_thread_env(connection->jvm);
     if (env == NULL) {
         /* If we can't get an environment, then the JVM is probably shutting down.  Don't crash. */
         return NULL;
     }
 
     struct aws_allocator *allocator = aws_jni_get_allocator();
     struct mqtt_jni_async_callback *callback = aws_mem_calloc(allocator, 1, sizeof(struct mqtt_jni_async_callback));
-    if (!callback) {
-        /* caller will throw when they get a null */
-        return NULL;
-    }
+    if (callback != NULL) {
+        callback->connection = connection;
+        callback->async_callback = async_callback ? (*env)->NewGlobalRef(env, async_callback) : NULL;
 
-    callback->connection = connection;
-    callback->async_callback = async_callback ? (*env)->NewGlobalRef(env, async_callback) : NULL;
+        aws_byte_buf_init(&callback->buffer, aws_jni_get_allocator(), 0);
+    }
 
-    aws_byte_buf_init(&callback->buffer, aws_jni_get_allocator(), 0);
+    aws_jni_release_thread_env(connection->jvm, env);
+    /********** JNI ENV RELEASE **********/
 
     return callback;
 }
 
 static void mqtt_jni_async_callback_destroy(struct mqtt_jni_async_callback *callback) {
     AWS_FATAL_ASSERT(callback && callback->connection);
-    JNIEnv *env = aws_jni_get_thread_env(callback->connection->jvm);
+
+    /********** JNI ENV ACQUIRE **********/
+    JNIEnv *env = aws_jni_acquire_thread_env(callback->connection->jvm);
```

Comment:
`mqtt_jni_async_callback_destroy()` is a helper function, not a callback, we could just pass `env` in as an argument